hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/artifacts/upload/AbstractFileTransferHandler.java

```diff
@@ -180,59 +193,89 @@ protected void assertStateConsistency(final FileUploadId fileUploadId, final Str
-        final OutputStream out = new FileOutputStream(tempFile);
-
-        tempFilePath = tempFile.getAbsolutePath();
-
-        return out;
     }
 
-    protected String getTempFilePath() {
-        return tempFilePath;
+    protected static void tryToCloseIOStream(final InputStream inputStream) {
+        try {
+            inputStream.close();
+        } catch (final IOException e1) {
+            LOG.error("Closing input stream caused an exception {}", e1);
+        }
     }
 
-    // Exception squid:S3655 - Optional access is checked in
-    // checkIfArtifactDetailsDispalyed subroutine
-    @SuppressWarnings("squid:S3655")
-    protected void transferArtifactToRepository(final FileUploadId fileUploadId, final long fileSize,
-            final String mimeType, final String tempFilePath) {
-
-        final File newFile = new File(tempFilePath);
-
-        final String filename = fileUploadId.getFilename();
-        LOG.info("Transfering tempfile {} - {} to repository", filename, tempFilePath);
-        try (FileInputStream fis = new FileInputStream(newFile)) {
+    private final class TransferArtifactToRepositoryRunnable implements Runnable {
+        private final InputStream inputStream;
+        private final FileUploadId fileUploadId;
+        private final String mimeType;
+        private final UI vaadinUi;
+
+        public TransferArtifactToRepositoryRunnable(final InputStream inputStream, final FileUploadId fileUploadId,
+                final String mimeType, final UI vaadinUi) {
+            this.inputStream = inputStream;
+            this.fileUploadId = fileUploadId;
+            this.mimeType = mimeType;
+            this.vaadinUi = vaadinUi;
+        }
 
-            artifactManagement.create(fis, fileUploadId.getSoftwareModuleId(), filename, null, null, true, mimeType,
-                    fileSize);
+        @Override
+        public void run() {
+            try {
+                UI.setCurrent(vaadinUi);
+                streamToRepository();
+            } catch (final RuntimeException e) {
+                publishUploadFailedEvent(fileUploadId, i18n.getMessage("message.upload.failed"), e);
+                tryToCloseIOStream(inputStream);
+                LOG.error("Failed to transfer file to repository", e);
+            }
+        }
 
-            publishUploadSucceeded(fileUploadId, fileSize, tempFilePath);
+        private void streamToRepository() {
+            if (fileUploadId == null) {
+                throw new ArtifactUploadFailedException();
+            }
 
-            eventBus.publish(this, new SoftwareModuleEvent(SoftwareModuleEventType.ARTIFACTS_CHANGED,
-                    fileUploadId.getSoftwareModuleId()));
+            final String filename = fileUploadId.getFilename();
+            LOG.info("Transfering file {} directly to repository", filename);
+            final Artifact artifact = uploadArtifact(filename).orElseThrow(ArtifactUploadFailedException::new);
```

Comment:
The method uploadArtifact returns an empty Optional in case of an error and then the calling code maps it back to an exception?

Reply:
It also returns an empty optional if the create method fails and returns null without throwing an exception. uploadArtifact is just a wrapper for the create methode, dealing with all possible upload failures.